scale/storage/brokers/host_broker.py

```diff
@@ -62,6 +61,62 @@ def get_file_system_paths(self, volume_path, files):
             paths.append(os.path.join(volume_path, scale_file.file_path))
         return paths
 
+    def list_files(self, volume_path, recursive, callback):
+        """See :meth:`storage.brokers.broker.Broker.list_files`
+        """
+
+        # List used to store files that are being returned on completion. Only used
+        # if callback is not defined or fails.
+        file_list = []
+
+        # List used to store files being batched. Flushed to callback on batch_size.
+        file_batch = []
+
+        batch_size = 1000
+
+        # Handle a full recursive walk of the directory tree.
+        if recursive:
+            for root, dirs, files in os.walk(volume_path):
+                for name in files:
+                    file_name = os.path.join(root, name)
+                    if os.path.isfile(file_name):
+                        file_details = FileDetails(file_name, os.path.getsize(file_name))
+                        file_batch.append(file_details)
+                        if len(file_batch) % batch_size == 0:
+                            try:
+                                if callback:
+                                    callback(file_batch)
+                                    file_batch = []
+                            except:
+                                logger.exception('list_files callback failure.')
+                            file_list.extend(file_batch)
+        # Handle identifying files only from a single directory.
+        else:
+            for result in os.listdir(volume_path):
```

Comment:
This has a lot of duplicate code from above. Recommend factoring that out into its own method.